### PR TITLE
Allow for cooperative release of funds

### LINKED ISSUE #693: [1 XMR] Allow for cooperative release of funds
The protocol allows for Alice to punish Bob by taking the BTC in case Bob never refunds. This can happen unintentionally as well. See https://github.com/comit-network/xmr-btc-swap/issues/688 for example.

In that case, the two parties would have agreed to a cooperative release of funds. This should be possible to implement if we add:

- A command to the `asb` to print the Monero secret share
- A command to the `cli` to claim the Monero using the secret share

### PR DESCRIPTION
**Pull request by patrini32**
Should close #693. Introduces new logic for the BtcPunished state.

### CODE REVIEW

**Comment by patrini32** [comment]
Oh, latest commit broke the code.

File: swap/src/protocol/bob/state.rs
Comment (patrini32) [inline swap/src/protocol/bob/state.rs:1]: I added this function to State3 because it's possible that Bob's database is in incorrect state.

**Comment by binarybaron** [comment]
@patrini32 Can you give @einliterflasche access to your fork of this repository? He has some improvements for this PR and that way he'll be able to push directly to your branch.

**Comment by patrini32** [comment]
> @patrini32 Can you give @Einliterflasche access to your fork of this repository? He has some improvements for this PR and that way he'll be able to push directly to your branch.

Sure.

**Comment by binarybaron** [comment]
> > @patrini32 Can you give @Einliterflasche access to your fork of this repository? He has some improvements for this PR and that way he'll be able to push directly to your branch.
> 
> Sure.

thanks :)

**Comment by patrini32** [comment]
Sent invite

**Comment by Einliterflasche** [comment]
thanks! I think using `let else` really pays off here

**Comment by patrini32** [comment]
> thanks! I think using `let else` really pays off here

Thank you for refactoring it.

File: swap/src/api/request.rs [hunk 1]
```diff
@@ -810,6 +816,159 @@ impl Request {
...
+                            .as_ref()
+                            .context("Could not get Monero wallet")?,
+                    ),
+                    context.config.env_config,
+                    event_loop_handle,
+                    monero_receive_address,
+                )
+                .await?;
+                tokio::spawn(event_loop.run().in_current_span());
+                tracing::info!("Asking Alice to reveal xmr key to us.");
```
Comment (binarybaron) [inline swap/src/api/request.rs:901]: nitpick: Remove the dot at the end

Comment (patrini32) [reply swap/src/api/request.rs:901]: Okay

File: swap/src/api/request.rs [hunk 2]
```diff
@@ -810,6 +816,159 @@ impl Request {
...
+                let response = swap
+                    .event_loop_handle
+                    .request_cooperative_xmr_redeem(swap_id)
+                    .await;
+                match response {
+                    Ok(response) => {
+                        tracing::info!("Alice revealed xmr key to us, redeeming xmr...");
```
Comment (binarybaron) [inline swap/src/api/request.rs:908]: same here xmr -> XMR

**Review by binarybaron — CHANGES_REQUESTED** [review]
This looks pretty good! I commented two small nitpicks. And two things we need to get done before we merge this:
1. Alice should return a proper error to Bob if she cannot provide the secret key (@Einliterflasche is working on this)
2. We want to add logic to the Bobs state machine for him to automatically try to cooperative redeem if he ends up in BtcPunished

**Comment by patrini32** [comment]
> This looks pretty good! I commented two small nitpicks. And two things we need to get done before we merge this:
> 
> 1. Alice should return a proper error to Bob if she cannot provide the secret key (@Einliterflasche is working on this)
> 2. We want to add logic to the Bobs state machine for him to automatically try to cooperative redeem if he ends up in BtcPunished

I implemented 2 initially, but it required changes to database.

**Comment by patrini32** [comment]
> This looks pretty good! I commented two small nitpicks. And three things we need to get done before we merge this:
> 
> 1. Alice should return a proper error to Bob if she cannot provide the secret key (@Einliterflasche is working on this)
> 2. We want to add logic to the Bobs state machine for him to automatically try to cooperative redeem if he ends up in BtcPunished
> 3. Think about adding the secret key to the BtcPunished state of Alice and also add the data needed for Bob to his states such that we dont need to iterate through all the states to find the data we need. If we do this, we should make this backwards compatible by possibly creating a migration script (add data all the required data to the states if they were created before the update). I think it's if we think about which option is best before implementing it.

I implemented 2, but it requires changes to multiple BobStates, because state3 doesn't contain monero_wallet_restore_blockheight, so we can't just change State6.

**Comment by patrini32** [comment]
Hmm, actually, what if pass monero_wallet_restore_blockheight to cancel?

**Comment by patrini32** [comment]
> This looks pretty good! I commented two small nitpicks. And three things we need to get done before we merge this:
> 
> 1. Alice should return a proper error to Bob if she cannot provide the secret key (@Einliterflasche is working on this)
> 2. We want to add logic to the Bobs state machine for him to automatically try to cooperative redeem if he ends up in BtcPunished
> 3. Think about adding the secret key to the BtcPunished state of Alice and also add the data needed for Bob to his states such that we dont need to iterate through all the states to find the data we need. If we do this, we should make this backwards compatible by possibly creating a migration script (add data all the required data to the states if they were created before the update). I think it's if we think about which option is best before implementing it.

I implemented 2 and added new test for it.
@binarybaron

**Comment by patrini32** [comment]
> This looks pretty good! I commented two small nitpicks. And three things we need to get done before we merge this:
> 
> 1. Alice should return a proper error to Bob if she cannot provide the secret key (@Einliterflasche is working on this)
> 2. We want to add logic to the Bobs state machine for him to automatically try to cooperative redeem if he ends up in BtcPunished
> 3. Think about adding the secret key to the BtcPunished state of Alice and also add the data needed for Bob to his states such that we dont need to iterate through all the states to find the data we need. If we do this, we should make this backwards compatible by possibly creating a migration script (add data all the required data to the states if they were created before the update). I think it's if we think about which option is best before implementing it.

Implemented all of these, can you review, please? @binarybaron

File: migrate/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    /// Input file
+    #[structopt(parse(from_os_str))]
+    file_name: PathBuf,
+}
+#[tokio::main(flavor = "current_thread")]
+async fn main() -> anyhow::Result<()> {
+    let args = Args::from_args();
+    let path_str = &args.file_name.display();
+    let pool = SqlitePool::connect(&format!("sqlite:{}", path_str)).await?;
+    println!("Copying {path_str} to {path_str}.original...");
```
Comment (patrini32) [inline migrate/src/main.rs:22]: I don't think that we need tracing there.

**Comment by patrini32** [comment]
@binarybaron, is it okay to use sql migrations?

**Comment by patrini32** [comment]
> > @binarybaron, is it okay to use sql migrations?
> 
> Yes that's fine. I'll have to review the SQL in depth. Can you add comments to the SQL query?

Sure

**Comment by patrini32** [comment]
> > @binarybaron, is it okay to use sql migrations?
> 
> Yes that's fine. I'll have to review the SQL in depth. Can you add comments to the SQL query?

Should I add comments to all SQL queries or just the first two? (Others are copies of these).

**Review by binarybaron — CHANGES_REQUESTED** [review]

File: swap/src/protocol/bob/swap.rs [hunk 4]
```diff
@@ -39,10 +37,12 @@ pub async fn run_until(
             swap.monero_receive_address,
         )
         .await?;
-
         swap.db
             .insert_latest_state(swap.id, current_state.clone().into())
             .await?;
+        if matches!(current_state, BobState::BtcPunished { .. }) {
```
Comment (patrini32) [inline swap/src/protocol/bob/swap.rs:43]: Is it okay to add this to run_until or should this function be more general?

**Comment by binarybaron** [comment]
@delta1 Do you have time to give this a review?

**Comment by patrini32** [comment]
Oh, sorry, @binarybaron, can you rerun the CI, I fixed tests.

**Comment by patrini32** [comment]
> Oh, sorry, @binarybaron, can you rerun the CI? I fixed tests.
Uh, that's weird. I ran this test several times and it didn't fail.

**Comment by patrini32** [comment]
@binarybaron, can you run CI again, I think there's some nondeterminism there.

**Comment by binarybaron** [comment]
Can you also add some more general comments to the SQL migration file? The ones you added are good but are very specific. They are quite hard to understand without having as much context.

For example: "This query adds the new state3 attribute that has been added in PR..."

**Comment by patrini32** [comment]
> Can you also add some more general comments to the SQL migration file? The ones you added are good but are very specific. They are quite hard to understand without having as much context.
> 
> For example: "This query adds the new state3 attribute that has been added in PR..."

Done.

**Comment by patrini32** [comment]
Are there any other changes I should make?

**Comment by patrini32** [comment]
> After letting a swap go to BtcPunished (with a CLI without these changes applied) and then running `swap --testnet history` gives an error even after the migrations have been applied.
> 
> ```
> ./target/debug/swap --testnet history
> Logging initialized to /Users/****/Library/Application Support/xmr-btc-swap/cli/testnet/logs
> Error: missing field `v` at line 1 column 3327
> ```

Hmm, that's odd, it worked when I tested it.

**Comment by binarybaron** [comment]
> > After letting a swap go to BtcPunished (with a CLI without these changes applied) and then running `swap --testnet history` gives an error even after the migrations have been applied.
> > ```
> > ./target/debug/swap --testnet history
> > Logging initialized to /Users/****/Library/Application Support/xmr-btc-swap/cli/testnet/logs
> > Error: missing field `v` at line 1 column 3327
> > ```
> 
> Hmm, that's odd, it worked when I tested it.

I'll investigate. Are you added the new v attribute to every state6 attribute in every state?

**Comment by patrini32** [comment]
> > > After letting a swap go to BtcPunished (with a CLI without these changes applied) and then running `swap --testnet history` gives an error even after the migrations have been applied.
> > > ```
> > > ./target/debug/swap --testnet history
> > > Logging initialized to /Users/****/Library/Application Support/xmr-btc-swap/cli/testnet/logs
> > > Error: missing field `v` at line 1 column 3327
> > > ```
> > 
> > 
> > Hmm, that's odd, it worked when I tested it.
> 
> I'll investigate. Are you added the new v attribute to every state6 attribute in every state?
Yes

**Comment by patrini32** [comment]
> > > After letting a swap go to BtcPunished (with a CLI without these changes applied) and then running `swap --testnet history` gives an error even after the migrations have been applied.
> > > ```
> > > ./target/debug/swap --testnet history
> > > Logging initialized to /Users/****/Library/Application Support/xmr-btc-swap/cli/testnet/logs
> > > Error: missing field `v` at line 1 column 3327
> > > ```
> > 
> > 
> > Hmm, that's odd, it worked when I tested it.
> 
> I'll investigate. Are you sure added the new v attribute to every state6 attribute in every state?

I tested it now. It works.

**Comment by binarybaron** [comment]
Can you use an enum for the errors instead of a string?

**Comment by patrini32** [comment]
> Can you use an enum for the errors instead of a string?

Do you want to parse errors on the cli side?

**Comment by binarybaron** [comment]
> > Can you use an enum for the errors instead of a string?
> 
> Do you want to parse errors on the cli side?

You can use `thiserror`

For example see how it is being used in `swap/src/network/swap_setup/alice.rs`

```rust
#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("ASB is running in resume-only mode")]
    ResumeOnlyMode,
    #[error("Amount {buy} below minimum {min}")]
    AmountBelowMinimum {
        min: bitcoin::Amount,
        buy: bitcoin::Amount,
    },
    #[error("Amount {buy} above maximum {max}")]
    AmountAboveMaximum {
        max: bitcoin::Amount,
        buy: bitcoin::Amount,
    },
    #[error("Unlocked balance ({balance}) too low to fulfill swapping {buy}")]
    BalanceTooLow {
        balance: monero_rpc::wallet::GetBalance,
        buy: bitcoin::Amount,
    },
    #[error("Failed to fetch latest rate")]
    LatestRateFetchFailed(#[source] Box<dyn std::error::Error + Send + Sync + 'static>),
    #[error("Failed to calculate quote")]
    SellQuoteCalculationFailed(#[source] anyhow::Error),
    #[error("Blockchain networks did not match, we are on {asb:?}, but request from {cli:?}")]
    BlockchainNetworkMismatch {
        cli: BlockchainNetwork,
        asb: BlockchainNetwork,
    },
}
```

File: swap/migrations/20240615140942_btcpunished_update.sql [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            json_object(
+                'state', -- {"Bob":{"BtcPunished":{"state": {}}}
+                json_insert(
+                    ( -- object that we insert properties into (original state6 from BtcCancelled state)
+                        SELECT json_extract(states.state, '$.Bob.BtcCancelled') -- Get state6 from BtcCancelled state.
+                        FROM swap_states AS states
+                        WHERE
+                            states.swap_id = swap_states.swap_id
+                            AND json_extract(
+                                states.state, '$.Bob.BtcCancelled'
```
Comment (binarybaron) [inline swap/migrations/20240615140942_btcpunished_update.sql:34]: We can enter the `BtcPunished` state without ever entering the `BtcCancelled` state. Can you extract this data from somewhere else?

**Review by binarybaron — CHANGES_REQUESTED** [review]
sql query

**Comment by binarybaron** [comment]
@Einliterflasche @delta1 Can you give this another glance before I merge this?

**Comment by delta1** [comment]
@binarybaron i will review tomorrow

**Review by delta1 — APPROVED** [review]
LGTM

**Pull request merged**